tracing-subscriber/src/subscribe/layered.rs

```diff
@@ -0,0 +1,476 @@
+            //
+            // Yes, this rule *is* slightly counter-intuitive, but it's
+            // necessary due to a weird edge case that can occur when two
+            // `Layered`s where one side is per-subscriber filtered and the other
+            // isn't are `Layered` together to form a tree. If we didn't have
+            // this rule, we would actually end up *ignoring* `Interest`s from
+            // the non-per-subscriber-filtered subscribers, since both branches would
+            // claim to have PLF.
+            //
+            // If you don't understand this...that's fine, just don't mess with
+            // it. :)
+            id if filter::is_plf_downcast_marker(id) => self
+                .subscriber
+                .downcast_raw(id)
+                .and(self.inner.downcast_raw(id)),
+
+            // Otherwise, try to downcast both branches normally...
+            _ => self
+                .subscriber
+                .downcast_raw(id)
+                .or_else(|| self.inner.downcast_raw(id)),
+        }
+    }
+}
+
+impl<'a, S, C> LookupSpan<'a> for Layered<S, C>
+where
+    C: Collect + LookupSpan<'a>,
+{
+    type Data = C::Data;
+
+    fn span_data(&'a self, id: &span::Id) -> Option<Self::Data> {
+        self.inner.span_data(id)
+    }
+
+    #[cfg(all(feature = "registry", feature = "std"))]
+    fn register_filter(&mut self) -> FilterId {
+        self.inner.register_filter()
+    }
+}
+
+impl<S, C> Layered<S, C>
+where
+    C: Collect,
+{
+    fn ctx(&self) -> Context<'_, C> {
+        Context::new(&self.inner)
+    }
+}
+
+impl<A, B, C> Layered<A, B, C>
+where
+    A: Subscribe<C>,
+    C: Collect,
+{
+    pub(super) fn new(subscriber: A, inner: B, inner_has_subscriber_filter: bool) -> Self {
+        #[cfg(all(feature = "registry", feature = "std"))]
+        let inner_is_registry = TypeId::of::<C>() == TypeId::of::<crate::registry::Registry>();
+        #[cfg(not(all(feature = "registry", feature = "std")))]
+        let inner_is_registry = false;
+
+        let inner_has_subscriber_filter = inner_has_subscriber_filter || inner_is_registry;
+        let has_subscriber_filter = filter::subscriber_has_plf(&subscriber);
+        Self {
+            subscriber,
+            inner,
+            has_subscriber_filter,
+            inner_has_subscriber_filter,
+            inner_is_registry,
+            _s: PhantomData,
+        }
+    }
+
+    fn pick_interest(&self, outer: Interest, inner: impl FnOnce() -> Interest) -> Interest {
+        if self.has_subscriber_filter {
+            return inner();
+        }
+
+        // If the outer subscriber has disabled the callsite, return now so that
+        // the inner subscriber/subscriber doesn't get its hopes up.
+        if outer.is_never() {
+            // If per-layer filters are in use, and we are short-circuiting
+            // (rather than calling into the inner type), clear the current
+            // per-layer filter interest state.
```

Comment:
```suggestion
            // If per-subscriber filters are in use, and we are short-circuiting
            // (rather than calling into the inner type), clear the current
            // per-subscriber filter interest state.
```